### PR TITLE
sql: add numeric data type formatting functions

### LINKED ISSUE #3781: sql: Add Data Type Formatting Functions
Eventually we are going to want to add support for converting various data types to formatted strings, and from formatted strings back again. Postgres handles this well and provides a lot of detail on the specifics of its implementation [here](http://www.postgresql.org/docs/8.1/static/functions-formatting.html).

- [ ] to_char(timestamp, text)
- [ ] to_char(interval, text)
- [ ] to_char(int, text)
- [ ] to_char(double precision, text)
- [ ] to_char(numeric, text)
- [ ] to_date(text, text)
- [ ] to_timestamp(text, text)
- [ ] to_timestamp(double precision)
- [ ] to_number(text, text)

### PR DESCRIPTION
**Pull request by rafiss**
### sql: add numeric data type formatting functions

Add PostgreSQL-compatible numeric formatting functions:
- to_char(int, text), to_char(float8, text), to_char(numeric, text)
- to_number(text, text)

These complement the existing datetime to_char variants. The implementation follows PostgreSQL's NUM_processor architecture with a separate keyword set and format parser for numeric patterns.

Supported format patterns: 9, 0, . (decimal), , (group separator), D, G, FM (fill mode), S, MI, PL, SG, PR (sign modes), L (currency), RN/rn (Roman numerals), TH/th (ordinal suffix), V (shift), EEEE (scientific notation), and B (blank zero).

All numeric overloads convert to apd.Decimal first and share a single formatting code path. The existing FormatCache is extended with prefixed keys ("dch:" / "num:") to avoid collisions between datetime and numeric format entries.

Release note (sql change): Added PostgreSQL-compatible numeric formatting functions to_char(int, text), to_char(float, text), to_char(numeric, text), and to_number(text, text). These functions format numbers as strings and parse formatted strings back to numbers using format patterns like '999,999.99', 'FM999.90', '9.999EEEE', and 'RN'.

---

### sql: add to_date and to_timestamp datetime parsing functions 

Add to_date(text, text) and to_timestamp(text, text) builtins that
parse date and timestamp values from formatted strings using DCH
(datetime character) format patterns. These are the reverse of the
existing to_char functions for timestamps.

The implementation supports the full set of PostgreSQL format patterns
including YYYY, MM, DD, HH24/HH12, MI, SS, MONTH/MON, AM/PM, BC/AD,
Julian day (J), day-of-year (DDD), ISO week dates (IYYY/IW/ID),
century (CC), Roman numeral months (RM), seconds since midnight (SSSS),
fractional seconds (MS/US/FF1-FF6), timezone offsets (TZH/TZM), and
the Y,YYY comma-separated year format. TZ/tz named timezone patterns
return an unsupported error, matching PostgreSQL behavior.

Also regenerates test stubs for the numeric_formatting logic test from
the previous commit.

Release note (sql change): Added to_date(text, text) and
to_timestamp(text, text) SQL functions that parse dates and timestamps
from formatted strings using PostgreSQL-compatible format patterns. For
example, to_date('2023-03-15', 'YYYY-MM-DD') returns a date, and
to_timestamp('2023-03-15 14:30:45', 'YYYY-MM-DD HH24:MI:SS') returns a
timestamptz.

---

Resolves: #3781

### CODE REVIEW

**Review by spilchen — CHANGES_REQUESTED** [review]
it'll be nice to get this long standing issue in

> @spilchen made 11 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-:-OmsuwHO4R0p2MpUUQIG:b-zedin6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on rafiss).

___
*[`pkg/util/tochar/dch_fromchar.go` line 141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omskl920h4A75x2-XiH:-Omskl920h4A75x2-XiI:b-duebmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L141)):*
> ```Go
> 		return 0, 0, pgerror.Newf(
> 			pgcode.InvalidDatetimeFormat,
> 			"invalid value \"%s\" for \"%d\"",
> ```

for the second arg, why do we include the length of `s`? Would it be more helpful to report the format specifier that was invalid?
___
*[`pkg/util/tochar/dch_fromchar.go` line 691 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsmSfp2hHJWxNS9ICB:-OmsmSfq26nJ3-3yMlkk:b-1y06av) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L691)):*
> ```Go
> 
> 			case DCH_DDD, DCH_ddd:
> 				val, n, err := fromCharParseIntLen(s, pos, 3, fm)
> ```

do we need to do any validation that the number produced is within a year? i.e. is this valid `SELECT to_date('999', 'DDD')`?
___
*[`pkg/util/tochar/dch_fromchar.go` line 856 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmssNAO9gD329Ff9K7e:-OmssNAO9gD329Ff9K7f:b-217w2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L856)):*
> ```Go
> 				pos = skipTHth(s, pos, fn.suffix)
> 
> 			case DCH_tzh:
> ```

is this any different from the `case DCH_TZH` branch? Can we merge the two?
___
*[`pkg/util/tochar/dch_fromchar.go` line 935 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmstYGr91dRtz40upNm:-OmstYGr91dRtz40upNn:b-wcxkwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L935)):*
> ```Go
> 	if tmfc.clock == 12 && tmfc.pm != fromCharFieldUnset {
> 		if tmfc.hh != fromCharFieldUnset {
> 			if hour < 1 || hour > 12 {
> ```

should the hour validation occur even if AM/PM isn't specified? i.e. `SELECT to_timestamp('13', 'HH12');`
___
*[`pkg/util/tochar/dch_fromchar.go` line 1103 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omspg4460Qr3JLFxgT7:-Omspg4460Qr3JLFxgT8:b3gcusm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L1103)):*
> ```Go
> 
> // decodeRune decodes a rune at position pos in string s.
> func decodeRune(s string, pos int) (rune, int) {
> ```

is the logic here any different from the standard library `utf8.DecodeRuneInString`?
___
*[`pkg/util/tochar/num_fromchar.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsr4CM84ae-E9ixQ6p:-Omsr4CN0yz7T5IFNjQB:b909hvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L104)):*
> ```Go
> 						b[0] = '-'
> 						numBuilder.Reset()
> 						numBuilder.Write(b)
> ```

nit: we set the sign quite a few times like this. It may warrant a helper to reduce the code duplication.



_Code quote:_
```Go
						b := []byte(numBuilder.String())
						b[0] = '-'
						numBuilder.Reset()
						numBuilder.Write(b)
```
___
*[`pkg/util/tochar/num_fromchar.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsudI_7kr8NPKNK_73:-OmsudI_7kr8NPKNK_74:bz7m7oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L107)):*
> ```Go
> 						inIdx++
> 					} else if inIdx < len(input) && input[inIdx] == '+' {
> 						b := []byte(numBuilder.String())
> ```

rewriting the string builder for sign seems like it defeats the purpose of the string builder. Is this efficient wrt memory allocation when we reset the builder?
___
*[`pkg/util/tochar/num_fromchar.go` line 162 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsk9bd1djvr-8sg8zX:-Omsk9bd1djvr-8sg8zY:b-3kav9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L162)):*
> ```Go
> 					if inIdx < len(input) && (input[inIdx] == '-' || input[inIdx] == '+') {
> 						b := []byte(numBuilder.String())
> 						b[0] = input[inIdx]
> ```

should we increment `inIdx` for this case?
___
*[`pkg/util/tochar/num_tochar.go` line 364 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsp6l863cC6O488nB5:-Omsp6l863cC6O488nB6:bwes9bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_tochar.go#L364)):*
> ```Go
> 					if lastRelevant < 0 || number[lastRelevant] != '.' {
> 						if n.key.id == NUM_D || n.key.id == NUM_d {
> 							sb.WriteByte('.')
> ```

nit: the `else` branch is the same as this. do we need the `if` at all?
___
*[`pkg/util/tochar/tochar.go` line 351 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsjecbCNwvUpmgwhn1:-OmsjecbCNwvUpmgwhn2:b-vdlgs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/tochar.go#L351)):*
> ```Go
> 			switch fn.key.id {
> 			case DCH_DAY:
> 				d = strings.ToLower(d)
> ```

not a problem with your change, but it had me looking at this code, shouldn't this be `strings.ToUpper()`?

File: pkg/util/tochar/num_fromchar.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+		return nil, pgerror.Newf(
+			pgcode.InvalidTextRepresentation,
+			"invalid input syntax for type numeric: %q", numStr,
+		)
+	}
+
+	// Apply V multiplier: divide by 10^multi.
+	if desc.isMulti() {
+		divisor := new(apd.Decimal)
+		divisor.SetFinite(1, int32(-desc.multi))
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/util/tochar/num_fromchar.go]: done

**Review by rafiss — COMMENTED** [review]
> @rafiss made 11 comments and resolved 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on spilchen).

___
*[`pkg/util/tochar/dch_fromchar.go` line 141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omskl920h4A75x2-XiH:-OmuwHaABTpJpCkV5PhE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L141)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

for the second arg, why do we include the length of `s`? Would it be more helpful to report the format specifier that was invalid?
</blockquote></details>

done
___
*[`pkg/util/tochar/dch_fromchar.go` line 691 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsmSfp2hHJWxNS9ICB:-OmuwJb5ClP7wgXNTIp6:bx9yv6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L691)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

do we need to do any validation that the number produced is within a year? i.e. is this valid `SELECT to_date('999', 'DDD')`?
</blockquote></details>

done. `SELECT to_date('999', 'DDD')` and `SELECT to_date('2023 366', 'YYYY DDD')` both fail in PG with "date/time field value
  out of range". I added validation that checks DDD against 365/366 depending on leap year, and added logic tests for the boundary cases.
___
*[`pkg/util/tochar/dch_fromchar.go` line 856 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmssNAO9gD329Ff9K7e:-OmuwLrR2Wqz3Wo1ppff:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L856)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

is this any different from the `case DCH_TZH` branch? Can we merge the two?
</blockquote></details>

done
___
*[`pkg/util/tochar/dch_fromchar.go` line 935 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmstYGr91dRtz40upNm:-OmuwMk85kzorgkTcBr0:b-8ugbuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L935)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

should the hour validation occur even if AM/PM isn't specified? i.e. `SELECT to_timestamp('13', 'HH12');`
</blockquote></details>

done; i added a test for this too
___
*[`pkg/util/tochar/dch_fromchar.go` line 1103 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omspg4460Qr3JLFxgT7:-OmuwNsV17RMGfyGMoR0:bcqvtfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/dch_fromchar.go#L1103)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

is the logic here any different from the standard library `utf8.DecodeRuneInString`?
</blockquote></details>

done; nice catch
___
*[`pkg/util/tochar/num_fromchar.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsr4CM84ae-E9ixQ6p:-OmuwPxP2b56d4Xr3DMd:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L104)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: we set the sign quite a few times like this. It may warrant a helper to reduce the code duplication.
</blockquote></details>

done
___
*[`pkg/util/tochar/num_fromchar.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsudI_7kr8NPKNK_73:-Omv2QSHBkHZNMSRO7Gu:b-xdqgpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L107)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

rewriting the string builder for sign seems like it defeats the purpose of the string builder. Is this efficient wrt memory allocation when we reset the builder?
</blockquote></details>

Rather than a helper that still rewrites the builder, I changed the approach entirely: the sign is now tracked as a separate byte variable and only prepended to the result at the end. This eliminates all the `[]byte` conversions / `Reset()` calls
___
*[`pkg/util/tochar/num_fromchar.go` line 162 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsk9bd1djvr-8sg8zX:-Omv2Y0SEcsKnP4RpJV1:bcqvtfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L162)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

should we increment `inIdx` for this case?
</blockquote></details>

done; nice catch
___
*[`pkg/util/tochar/num_tochar.go` line 364 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-Omsp6l863cC6O488nB5:-Omv2bHp29ZFFbL_KPUe:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_tochar.go#L364)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: the `else` branch is the same as this. do we need the `if` at all?
</blockquote></details>

done
___
*[`pkg/util/tochar/tochar.go` line 351 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsjecbCNwvUpmgwhn1:-OmuwERZ7GBfLqk5NX1v:b-w62m3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfef390f285e380465740b3137d97d489f3b31b5/pkg/util/tochar/tochar.go#L351)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

not a problem with your change, but it had me looking at this code, shouldn't this be `strings.ToUpper()`?
</blockquote></details>

great catch. yes that's right

**Review by spilchen — COMMENTED** [review]
> @spilchen reviewed 3 files, made 1 comment, and resolved 6 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on rafiss).

___
*[`pkg/util/tochar/num_fromchar.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsudI_7kr8NPKNK_73:-Omxy7Yp0wwzqev6EYFN:b-3gqiyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L107)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Rather than a helper that still rewrites the builder, I changed the approach entirely: the sign is now tracked as a separate byte variable and only prepended to the result at the end. This eliminates all the `[]byte` conversions / `Reset()` calls
</blockquote></details>

Hmm, what you suggest is a good approach since it prevents memory reallocation within the buffer. But I still see the []byte conversion and Reset() calls in the `setBuilderSign` helper. What am I missing?

**Review by rafiss — COMMENTED** [review]
ready for another look. the latest changes have a quite a few additional fixes. these were all driven by a new fuzz test that compares this implementation to Postgres. I used Claude to write the code for generating random, representative inputs, then collect all the test cases that failed. Claude was also helpful to compare some obscure edge cases in the format/parsing logic with the Postgres source code to fix these bugs.

> @rafiss made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-:-OnAuyWo0CcvOlvNAjIv:bppaeh6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on spilchen).

___
*[`pkg/util/tochar/num_fromchar.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OmsudI_7kr8NPKNK_73:-OnAuoPg7l1ey1l-uBgd:b-ree4wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c777087586bf1169cc1227317ceae2357f6a6d5/pkg/util/tochar/num_fromchar.go#L107)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Hmm, what you suggest is a good approach since it prevents memory reallocation within the buffer. But I still see the []byte conversion and Reset() calls in the `setBuilderSign` helper. What am I missing?
</blockquote></details>

i've removed the setBuilderSign function entirely now. i may have made my last reply during an intermediate state.

**Review by spilchen — APPROVED** [review]
:lgtm: I like that you added the fuzz test.

> @spilchen reviewed 3 files and all commit messages, made 3 comments, and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-:-OnHXzA1-TGdFoWqXT-W:b2fh1xt)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on rafiss).

___
*[`pkg/util/tochar/num_fromchar.go` line 342 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OnHSN9BDxm9l7WAO-1v:-OnHSN9C0rJzzNl5iAby:brij6cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2f2d6678c4be8dc979952ac0f1e5184bb50de9/pkg/util/tochar/num_fromchar.go#L342)):*
> ```Go
> 
> // charLen returns the byte length of the first UTF-8 character in s.
> func charLen(s string) int {
> ```

nit: could we rewrite this function to use utf8.DecodeRuneInString?
```
_, size := utf8.DecodeRuneInString(s)
return size
```
___
*[`pkg/util/tochar/num_tochar.go` line 304 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OnHWwFN0OxOfgj3YtwI:-OnHWwFN0OxOfgj3YtwJ:b-slsjmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2f2d6678c4be8dc979952ac0f1e5184bb50de9/pkg/util/tochar/num_tochar.go#L304)):*
> ```Go
> 	if desc.isPlus() || desc.isMinus() {
> 		// MI/PL/SG - write sign in the pattern position, not before number.
> 		if desc.isPlus() && !desc.isMinus() {
> ```

nit: could we simplify this block to just be:
```
signWrote = desc.isMinus()
```

**Review by rafiss — COMMENTED** [review]
TFTR!

/trunk merge

> @rafiss made 3 comments and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-:-OnI0bwp9yZkai_N6MS0:b-80z54n)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on spilchen).

___
*[`pkg/util/tochar/num_fromchar.go` line 342 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OnHSN9BDxm9l7WAO-1v:-OnI0_6q2AObZJ2KzK13:bf9hgrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2f2d6678c4be8dc979952ac0f1e5184bb50de9/pkg/util/tochar/num_fromchar.go#L342)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: could we rewrite this function to use utf8.DecodeRuneInString?
```
_, size := utf8.DecodeRuneInString(s)
return size
```
</blockquote></details>

done; i just removed this function since there were only 2 usages.
___
*[`pkg/util/tochar/num_tochar.go` line 304 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/164672#-OnHWwFN0OxOfgj3YtwI:-OnI0ZPh76JtT6G_ba8l:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2f2d6678c4be8dc979952ac0f1e5184bb50de9/pkg/util/tochar/num_tochar.go#L304)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: could we simplify this block to just be:
```
signWrote = desc.isMinus()
```
</blockquote></details>

done
